core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -2815,9 +2823,99 @@ public synchronized void save() throws IOException {
      */
     @edu.umd.cs.findbugs.annotations.SuppressWarnings("ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD")
     public void cleanUp() {
-        for (ItemListener l : ItemListener.all())
-            l.onBeforeShutdown();
+        if (theInstance != this && theInstance != null) {
+            LOGGER.log(Level.WARNING, "This instance is no longer the singleton, ignoring cleanUp()");
+            return;
+        }
+        synchronized (Jenkins.class) {
+            if (cleanUpStarted) {
+                LOGGER.log(Level.WARNING, "Jenkins.cleanUp() already started, ignoring repeated cleanUp()");
+                return;
+            }
+            cleanUpStarted = true;
+        }
+        try {
+            LOGGER.log(Level.INFO, "Stopping Jenkins");
+
+            final List<Throwable> errors = new ArrayList<>();
+
+            fireBeforeShutdown(errors);
+
+            _cleanUpRunTerminators(errors);
+
+            terminating = true;
+
+            final Set<Future<?>> pending = _cleanUpDisconnectComputers(errors);
+
+            _cleanUpShutdownUDPBroadcast(errors);
+
+            _cleanUpCloseDNSMulticast(errors);
+
+            _cleanUpInterruptReloadThread(errors);
+
+            _cleanUpShutdownTriggers(errors);
+
+            _cleanUpShutdownTimer(errors);
+
+            _cleanUpShutdownTcpSlaveAgent(errors);
+
+            _cleanUpShutdownPluginManager(errors);
+
+            _cleanUpPersistQueue(errors);
+
+            _cleanUpShutdownThreadPoolForLoad(errors);
+
+            _cleanUpAwaitDisconnects(errors, pending);
+
+            _cleanUpPluginServletFilters(errors);
+
+            _cleanUpReleaseAllLoggers(errors);
+
+            LOGGER.log(Level.INFO, "Jenkins stopped");
+
+            if (!errors.isEmpty()) {
+                StringBuilder message = new StringBuilder("Unexpected issues encountered during cleanUp: ");
+                Iterator<Throwable> iterator = errors.iterator();
+                message.append(iterator.next().getMessage());
+                while (iterator.hasNext()) {
+                    message.append("; ");
+                    message.append(iterator.next().getMessage());
+                }
+                iterator = errors.iterator();
+                RuntimeException exception = new RuntimeException(message.toString(), iterator.next());
+                while (iterator.hasNext()) {
+                    exception.addSuppressed(iterator.next());
+                }
+                throw exception;
```

Comment:
This seems wrong. It means shutdown will hang if one of the terminators failed, whether that was important or not.

Reply:
> resulting in tests that fail to clean up correctly remaining hidden and consequently resulting in random test failures

sounds like this would better be activated iff `Main.isUnitTest`.